python/src/aiconfig/editor/server/server.py

```diff
@@ -49,137 +88,120 @@ def home():
-    res_aiconfig = path_val.and_then(safe_load_from_disk)
-    match res_aiconfig:
-        case Ok(aiconfig):
-            LOGGER.warning(f"Loaded AIConfig from {path_val}. This may have overwritten in-memory changes.")
-            state.aiconfig = aiconfig
-            return HttpResponseWithAIConfig(message="Loaded", aiconfig=aiconfig).to_flask_format()
-        case Err(e):
-            return HttpResponseWithAIConfig(message=f"Failed to load AIConfig: {path_val}, {e}", code=400, aiconfig=None).to_flask_format()
+    request_json = request.get_json()
+    path: str | None = request_json.get("path", None)
+    if path is None:
+        aiconfig = state.aiconfig
+        if aiconfig is None:
+            return HttpResponseWithAIConfig(message="No AIConfig loaded", code=400, aiconfig=None).to_flask_format()
+        else:
+            return HttpResponseWithAIConfig(message="AIConfig already loaded. Here it is!", aiconfig=aiconfig).to_flask_format()
+    else:
+        path_val = get_validated_path(path)
+        res_aiconfig = path_val.and_then(safe_load_from_disk)
+        match res_aiconfig:
+            case Ok(aiconfig):
+                LOGGER.warning(f"Loaded AIConfig from {path_val}. This may have overwritten in-memory changes.")
+                state.aiconfig = aiconfig
+                return HttpResponseWithAIConfig(message="Loaded", aiconfig=aiconfig).to_flask_format()
+            case Err(e):
+                return HttpResponseWithAIConfig(message=f"Failed to load AIConfig: {path_val}, {e}", code=400, aiconfig=None).to_flask_format()
 
 
 @app.route("/api/save", methods=["POST"])
 def save() -> FlaskResponse:
     state = get_server_state(app)
-    if state.aiconfig is None:
-        return HttpResponseWithAIConfig(message="No AIConfig in memory, nothing to save.", code=400, aiconfig=None).to_flask_format()
-    else:
-        aiconfig: AIConfigRuntime = state.aiconfig
-
-        path_val = _validated_request_path(allow_create=True)
-        res_save: Result[HttpResponseWithAIConfig, str] = result.do(
-            Ok(HttpResponseWithAIConfig(message="Saved to disk", aiconfig=aiconfig))
-            for path_val_ok in path_val
-            for _ in safe_save_to_disk(aiconfig, path_val_ok)
-        )
-        return res_save.unwrap_or(
-            HttpResponseWithAIConfig(
-                #
-                message="Failed to save to disk",
-                code=400,
-                aiconfig=None,
-            )
-        ).to_flask_format()
+    aiconfig = state.aiconfig
+    request_json = request.get_json()
+
+    def _op(aiconfig: AIConfigRuntime, op_args: OpArgs) -> Result[None, str]:
+        method_name = MethodName("save")
+        return safe_run_aiconfig_method(aiconfig, method_name, op_args)
+
+    path_val = _validated_request_path(request_json, allow_create=True)
+    op_args: Result[OpArgs, str] = result.do(
+        Ok(OpArgs({"path": path_val_ok}))
+        #
+        for path_val_ok in path_val
+    )
```

Comment:
What is this doing? Specifically, isn't path_val just a string path? I'm confusing by the for ... in

Reply:
`path_val` might actually be invalid: it's a Result[ValidatedPath, str] where the str contains the error message. This expression basically unboxes the result and returns the expression on 146 if it's valid, otherwise it short-circuits and returns the Err value. 

You can think of it as kind of like a Result comprehension, sort of similar to a list comprehension but for Result. 

The advantage is significant: I don't have to worry about exceptions, which are not checked statically even by strict type checking like Pyright. All Python exceptions are unchecked. Turning exceptions into Results allows Pyright to find unhandled errors, which really reduces debugging time in my experience. In practice, I'm finding that if my code type checks, it runs, and a large fraction of tricky bugs are cases where I had to suppress Pyright for one reason or another.